### ISSUE TITLE
GPU memory usage is not proportional to the image area.

### ISSUE THREAD

**Issue by chenyp132** [ultralytics/ultralytics#22713]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

When I predict with yolo11n, imgsz = 2000 and 4000, the GPU memory usage is 800M and 2390M respectively. However, when I use imgsz=8000, the GPU memory usage reaches 30000M, which is not proportional to the image area. Why does this happen?
Existing optimizations for large-size images all focus on memory efficiency, with no mention of the relevant issue.

### Additional

This is my environmen：
torch                     2.7.1
torchvision               0.22.1            
ultralytics               8.3.159
opencv-python             4.11.0.86
pandas                    1.4.4

**Comment by UltralyticsAssistant** [comment]
👋 Hello @chenyp132, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support as quickly as possible — an Ultralytics engineer will also review your issue and follow up soon.

From your description, this looks like a ❓ Question related to GPU memory behavior with large `imgsz` values in YOLO11. To help the team better understand and investigate, please provide as much detail as you can, such as:

- Exact command or Python code you are using for prediction  
- Model variant (`yolo11n` in your case) and any custom settings (e.g. `batch`, `device`, `half`, `workers`)  
- Whether you are using a single image or batches, and their dimensions  
- Screenshots or logs showing GPU memory usage for the different image sizes  

We also recommend a visit to the [Docs](https://docs.ultralytics.com/) where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered ✅.

If at any point you suspect this might actually be a 🐛 Bug, please also share a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) (MRE) so the team can reproduce the behavior on their side.

If this is a custom training ❓ Question (e.g. about training large images), please provide:

- Dataset image examples  
- Training command and logs  
- Confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)  

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Before the team dives deeper, please verify the issue on the latest version of `ultralytics`, as it may already be resolved there. Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a **Python>=3.8** environment with **PyTorch>=1.8**:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

If you can reproduce the GPU memory behavior in one of these environments, that information will be very helpful for the engineers 🧠.

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Good observation—GPU memory in PyTorch/YOLO won’t scale perfectly with image area, especially once you reach very large sizes like `imgsz=8000`: the main cost does grow roughly with pixels, but CUDA/cuDNN also allocate extra temporary workspaces and PyTorch’s caching allocator can switch algorithms at certain size thresholds, which can cause sudden jumps like the one you see at 8K; for such resolutions it’s generally better to downscale or tile the image into smaller patches for inference instead of using a single very large `imgsz`, and for more context on GPU behavior you can also check the GPU-related notes in [Troubleshooting common YOLO issues](https://docs.ultralytics.com/guides/yolo-common-issues/).

**Comment by chenyp132** [comment]
Thank you for your reply and explanation. Yesterday, I disassembled and analyzed the parameters of each layer and found that the change in parameter count strictly increases in proportion to the area. I speculated that the massive increase in GPU memory was caused by changes at the GPU underlying logic, and your reply has helped me gain a deeper understanding of the underlying operating mechanisms of CUDA and PyTorch.


陈云鹏
***@***.***







        



         原始邮件
         
       
发件人：Glenn Jocher ***@***.***&gt;
发件时间：2025年11月19日 12:40
收件人：ultralytics/ultralytics ***@***.***&gt;
抄送：chenyp132 ***@***.***&gt;, State change ***@***.***&gt;
主题：Re: [ultralytics/ultralytics] GPU memory usage is not proportional tothe image area. (Issue #22713)



glenn-jocher left a comment (ultralytics/ultralytics#22713)

Good observation—GPU memory in PyTorch/YOLO won’t scale perfectly with image area, especially once you reach very large sizes like imgsz=8000: the main cost does grow roughly with pixels, but CUDA/cuDNN also allocate extra temporary workspaces and PyTorch’s caching allocator can switch algorithms at certain size thresholds, which can cause sudden jumps like the one you see at 8K; for such resolutions it’s generally better to downscale or tile the image into smaller patches for inference instead of using a single very large imgsz, and for more context on GPU behavior you can also check the GPU-related notes in Troubleshooting common YOLO issues.

**Comment by glenn-jocher** [comment]
Nice work digging into the layer-wise params—what you describe makes sense: parameter count covers only the persistent weights, while the big jump is coming from intermediate activations and backend workspaces; if you’d like to probe it further, you can compare memory per `imgsz` with something like:

```python
import torch
from ultralytics import YOLO

model = YOLO("yolo11n.pt").to("cuda")

for s in [2000, 4000, 8000]:
    x = torch.zeros(1, 3, s, s, device="cuda")
    torch.cuda.reset_peak_memory_stats()
    with torch.no_grad():
        model(x)
    print(s, "peak MB:", torch.cuda.max_memory_allocated() / 1e6)
```

which should give you a clearer view of where those step changes occur.
